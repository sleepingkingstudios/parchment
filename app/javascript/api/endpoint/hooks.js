import { useEffect } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';

const generateHooks = ({ performRequest, selector }) => {
  const useEndpoint = (fn = state => state) => useSelector(
    state => fn(selector(state)),
    shallowEqual,
  );

  const usePerformRequest = params => () => {
    const dispatch = useDispatch();

    useEffect(() => { performRequest(params)(dispatch); });
  };

  return {
    useEndpoint,
    usePerformRequest,
  };
};

export default generateHooks;
