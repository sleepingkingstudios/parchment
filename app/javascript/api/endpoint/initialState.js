import { INITIALIZED } from '../status';

const extractData = (options) => {
  if (typeof options === 'undefined' || options === null) { return {}; }

  const { data } = options;

  if (typeof data === 'undefined' || data === null) { return {}; }

  return data;
};

const generateInitialState = (options) => {
  const data = extractData(options);
  const errors = {};
  const status = INITIALIZED;

  return {
    data,
    errors,
    status,
  };
};

export default generateInitialState;
