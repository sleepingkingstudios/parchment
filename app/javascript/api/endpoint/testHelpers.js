/* eslint-env jest */

export const shouldGenerateTheEndpointActions = ({ actions, namespace }) => {
  const {
    REQUEST_FAILURE,
    REQUEST_PENDING,
    REQUEST_SUCCESS,
    requestFailure,
    requestPending,
    requestSuccess,
  } = actions;

  describe('REQUEST_FAILURE', () => {
    it('should define the action', () => {
      expect(REQUEST_FAILURE).toEqual(`${namespace}/requestFailure`);
    });
  });

  describe('REQUEST_PENDING', () => {
    it('should define the action', () => {
      expect(REQUEST_PENDING).toEqual(`${namespace}/requestPending`);
    });
  });

  describe('REQUEST_SUCCESS', () => {
    it('should define the action', () => {
      expect(REQUEST_SUCCESS).toEqual(`${namespace}/requestSuccess`);
    });
  });

  describe('requestFailure', () => {
    it('should be a function', () => {
      expect(typeof requestFailure).toEqual('function');
    });

    it('should create the action', () => {
      const errors = { widget: ['not_found'] };
      const action = requestFailure(errors);

      expect(action).toEqual({
        type: REQUEST_FAILURE,
        payload: { errors },
      });
    });
  });

  describe('requestPending', () => {
    it('should be a function', () => {
      expect(typeof requestPending).toEqual('function');
    });

    it('should create the action', () => {
      const action = requestPending();

      expect(action).toEqual({ type: REQUEST_PENDING, payload: {} });
    });
  });

  describe('requestSuccess', () => {
    it('should be a function', () => {
      expect(typeof requestSuccess).toEqual('function');
    });

    it('should create the action', () => {
      const data = {
        id: '00000000-0000-0000-0000-000000000000',
        name: 'Inigo Montoya',
      };
      const action = requestSuccess(data);

      expect(action).toEqual({
        type: REQUEST_SUCCESS,
        payload: { data },
      });
    });
  });
};

export const shouldGenerateTheSelector = ({ namespace, selector }) => {
  const generateState = (data) => {
    const segments = namespace.split('/');

    return segments.reverse().reduce((memo, key) => {
      const obj = {};

      obj[key] = memo;

      return obj;
    }, data);
  };

  it('should be a function', () => {
    expect(typeof selector).toEqual('function');
  });

  it('should select the data', () => {
    const data = { value: 'value' };
    const state = generateState(data);

    expect(selector(state)).toEqual(data);
  });
};
