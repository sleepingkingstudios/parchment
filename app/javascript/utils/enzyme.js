import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
