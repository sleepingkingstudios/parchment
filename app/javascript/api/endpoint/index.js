import defaultGenerateHooks from './hooks';
import defaultGenerateSelector from './selector';
import {
  applyMiddleware,
  selectMiddleware,
} from '../middleware/utils';
import { valueOrDefault } from '../../utils/object';

const applyMiddlewareToRequest = (request, middleware) => {
  request.handleFailure = applyMiddleware(
    request.handleFailure,
    selectMiddleware(middleware, 'handleFailure'),
  );
  request.handlePending = applyMiddleware(
    request.handlePending,
    selectMiddleware(middleware, 'handlePending'),
  );
  request.handleSuccess = applyMiddleware(
    request.handleSuccess,
    selectMiddleware(middleware, 'handleSuccess'),
  );

  return request;
};

class ApiEndpoint {
  constructor(options) {
    const {
      generateActions,
      generateHooks,
      generateInitialState,
      generateReducer,
      generateRequest,
      generateSelector,
      method,
      middleware,
      namespace,
      url,
    } = options;
    const actions = generateActions({ namespace });
    const data = valueOrDefault(options.data, {});
    const initialState = generateInitialState({ data });
    const reducer = generateReducer({ actions, initialState });
    const request = generateRequest({
      actions,
      method,
      namespace,
      url,
    });
    const { performRequest } = request;
    const selector = valueOrDefault(generateSelector, defaultGenerateSelector)({ namespace });
    const hooks = valueOrDefault(generateHooks, defaultGenerateHooks)({
      actions,
      performRequest,
      selector,
    });

    this.actions = actions;
    this.hooks = hooks;
    this.namespace = namespace;
    this.reducer = applyMiddleware(
      reducer,
      selectMiddleware(middleware, 'handleAction'),
    );
    this.request = applyMiddlewareToRequest(request, middleware);
    this.selector = selector;
  }
}

export default ApiEndpoint;
