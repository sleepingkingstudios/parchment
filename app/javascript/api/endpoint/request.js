import fetch from 'cross-fetch';
import { formatErrors } from '../../components/form/utils';
import {
  camelizeKeys,
  deepAccessProperty,
  underscoreKeys,
  valueOrDefault,
} from '../../utils/object';
import { interpolate } from '../../utils/string';

const getData = ({ namespace, state }) => {
  const segments = namespace.split('/');
  const raw = deepAccessProperty(state, 'data', segments);

  return underscoreKeys(raw);
};

const buildRequest = ({ getState, method, namespace }) => {
  const request = {
    method,
    headers: {
      'Content-Type': 'application/json',
    },
  };

  if (method === 'GET' || method === 'DELETE') { return request; }

  const state = getState();
  const data = getData({ namespace, state });

  return Object.assign(
    request,
    { body: JSON.stringify(data) },
  );
};

const extractErrors = json => valueOrDefault(
  deepAccessProperty(json, 'errors', ['error', 'data']), [],
);

const processResponse = async (response) => {
  const {
    headers,
    ok,
    status,
    statusText,
  } = response;
  const json = await response.json();

  return {
    headers,
    json,
    ok,
    status,
    statusText,
  };
};

class ApiRequest {
  constructor(options) {
    const {
      actions,
      method,
      namespace,
      url,
    } = options;

    this.method = method.toUpperCase();
    this.namespace = namespace;
    this.url = url;

    this.defineHandlers = this.defineHandlers.bind(this);
    this.performRequest = this.performRequest.bind(this);

    this.defineHandlers({ actions });
  }

  defineHandlers({ actions }) {
    const {
      requestPending,
      requestFailure,
      requestSuccess,
    } = actions;

    this.handleFailure = ({ dispatch, response }) => {
      const { json } = response;
      const errors = extractErrors(json);

      dispatch(requestFailure(formatErrors(errors)));
    };

    this.handlePending = ({ dispatch }) => {
      dispatch(requestPending());
    };

    this.handleSuccess = async ({ dispatch, response }) => {
      const { json } = response;

      dispatch(requestSuccess(camelizeKeys(json.data)));
    };
  }

  performRequest(params = {}) {
    const request = this;

    return async (dispatch, getState) => {
      const {
        handleFailure,
        handlePending,
        handleSuccess,
        method,
        namespace,
        url,
      } = request;

      handlePending({ dispatch, getState });

      const fullUrl = interpolate(url, /:(\w+)/g, params);
      const rawResponse = await fetch(
        fullUrl,
        buildRequest({
          getState,
          method,
          namespace,
        }),
      );
      const response = await processResponse(rawResponse);

      if (response.ok) {
        handleSuccess({ dispatch, getState, response });
      } else {
        handleFailure({ dispatch, getState, response });
      }
    };
  }
}

export default ApiRequest;
