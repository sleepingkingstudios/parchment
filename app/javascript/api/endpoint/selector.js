import { dig } from '../../utils/object';

const generateSelector = ({ namespace }) => state => dig(state, ...namespace.split('/'));

export default generateSelector;
