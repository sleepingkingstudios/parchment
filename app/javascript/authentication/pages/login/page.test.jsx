import React from 'react';
import { shallow } from 'enzyme';

import AuthenticationLoginPage from './page';
import endpoint from '../../store/loginForm';

describe('<AuthenticationLoginPage />', () => {
  const defaultProps = {};

  it('should render the page', () => {
    const breadcrumbs = [];
    const rendered = shallow(<AuthenticationLoginPage {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Page');
    expect(rendered).toHaveClassName('page-login');
    expect(rendered).toHaveProp({ breadcrumbs });
  });

  it('should render the heading', () => {
    const rendered = shallow(<AuthenticationLoginPage {...defaultProps} />);
    const heading = rendered.find('h1');

    expect(heading).toExist();
    expect(heading).toHaveText('Log In');
  });

  it('should render the form', () => {
    const rendered = shallow(<AuthenticationLoginPage {...defaultProps} />);
    const form = rendered.find('ConnectedAuthenticationLoginForm');

    expect(form).toExist();
    expect(form).toHaveProp({ endpoint });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<AuthenticationLoginPage {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });
});
