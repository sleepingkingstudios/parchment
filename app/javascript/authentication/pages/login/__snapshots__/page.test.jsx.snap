// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthenticationLoginPage /> should match the snapshot 1`] = `
<Page
  breadcrumbs={Array []}
  className="page-login"
  layout={null}
  showUser={true}
  subtitle="5e Campaign Companion"
  title="Parchment"
>
  <div>
    <h1>
      Log In
    </h1>
    <ConnectedAuthenticationLoginForm
      endpoint={
        FormEndpoint {
          "actions": Object {
            "REQUEST_FAILURE": "authentication/loginForm/requestFailure",
            "REQUEST_PENDING": "authentication/loginForm/requestPending",
            "REQUEST_SUCCESS": "authentication/loginForm/requestSuccess",
            "SET_FORM_DATA": "authentication/loginForm/setFormData",
            "UPDATE_FORM_DATA": "authentication/loginForm/updateFormData",
            "UPDATE_FORM_FIELD": "authentication/loginForm/updateFormField",
            "requestFailure": [Function],
            "requestPending": [Function],
            "requestSuccess": [Function],
            "setFormData": [Function],
            "updateFormData": [Function],
            "updateFormField": [Function],
          },
          "hooks": Object {
            "useEndpoint": [Function],
            "useSubmitForm": [Function],
            "useUpdateForm": [Function],
          },
          "namespace": "authentication/loginForm",
          "reducer": [Function],
          "request": FormRequest {
            "buildRequest": [Function],
            "defineHandlers": [Function],
            "handleFailure": [Function],
            "handlePending": [Function],
            "handleSuccess": [Function],
            "method": "POST",
            "namespace": "authentication/loginForm",
            "performRequest": [Function],
            "url": "/api/authentication/session",
          },
          "selector": [Function],
        }
      }
    />
  </div>
</Page>
`;
